core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -384,40 +355,62 @@ public CompletableFuture<Map<TopicIdPartition, ShareAcknowledgeResponseData.Part
     }
 
     public ShareFetchContext newContext(String groupId, Map<TopicIdPartition,
-            ShareFetchRequest.SharePartitionData> shareFetchData, List<TopicIdPartition> toForget,
-                                        Map<Uuid, String> topicNames, ShareFetchMetadata reqMetadata) {
+            ShareFetchRequest.SharePartitionData> shareFetchData, List<TopicIdPartition> toForget, ShareFetchMetadata reqMetadata) {
         ShareFetchContext context;
+        // TopicPartition with maxBytes as 0 should not be added in the cachedPartitions
+        Map<TopicIdPartition, ShareFetchRequest.SharePartitionData> shareFetchDataWithMaxBytes = new HashMap<>();
+        shareFetchData.forEach((tp, sharePartitionData) -> {
+            if (sharePartitionData.maxBytes > 0) shareFetchDataWithMaxBytes.put(tp, sharePartitionData);
+        });
         if (reqMetadata.isFull()) {
-            String removedFetchSessionStr = "";
             // TODO: We will handle the case of INVALID_MEMBER_ID once we have a clear definition for it
          //  if (!Objects.equals(reqMetadata.memberId(), ShareFetchMetadata.INVALID_MEMBER_ID)) {
             ShareSessionKey key = shareSessionKey(groupId, reqMetadata.memberId());
-            if (cache.remove(key) != null)
-                removedFetchSessionStr = "Removed share session with key " + key;
-
+            String removedFetchSessionStr = "";
             if (reqMetadata.epoch() == ShareFetchMetadata.FINAL_EPOCH) {
                 // If the epoch is FINAL_EPOCH, don't try to create a new session.
-                context = new FinalContext(shareFetchData);
+                if (!shareFetchDataWithMaxBytes.isEmpty()) {
+                    throw Errors.INVALID_REQUEST.exception();
+                }
+                context = new FinalContext();
+                if (cache.remove(key) != null) {
+                    removedFetchSessionStr = "Removed share session with key " + key;
+                    log.debug(removedFetchSessionStr);
+                }
             } else {
-                context = new ShareSessionContext(time, cache, reqMetadata, shareFetchData);
+                if (cache.remove(key) != null) {
```

Comment:
lines 376-378 and 381-383 are the same code lines in both if and else blocks, so can we move them to a common place?

Reply:
Thanks for the review ! Even when the lines are same, the order of execution is important here, if the if case, I want it to execute after creation of FinalContext but in the else case, I want to execute it before the creation of ShareSessionContext.